### PR TITLE
Convert transform math to C++ and move from Roadway to Sensor Fusion

### PR DESCRIPTION
**Pull request by msmcconnell**
# PR Details
Resolves Issue #88 
## Description
Prevents a race condition between sensor fusion and roadway described in the referenced issue. TF publication is moved to sensor fusion. The code is a direct port of the TransformMaintainer.java class from the roadway package into C++. This code was previously in the fix/sensor_fusion_lat_lon_conv branch of the pre-release repo, but was never merged in due to it being a lower priority issue. After having difficulties with detecting NCVs while testing the Traffic Signal Plugin one day, I moved the port. I have not observed an issues resulting from this change in the week or so since performing the port.

If this is merged in the documentation for roadway/sensor fusion will need to be updated. I will add a Jira ticket for that.  

## Related Issue
#88 

## Motivation and Context

Improve object detection and sensor fusion

## How Has This Been Tested?

C++ unit tests were created by porting the java tests and can be run with the catkin_make run_tests command. 

The code has been running during Traffic Signal Plugin tests for the last week to week and a half. 

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
[CARMA Contributing Guide](Contributing.md) 
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: sensor_fusion/src/sensor_fusion.h [hunk 1]
```diff
@@ -118,7 +125,12 @@ class SensorFusionApplication
 
     bool use_interface_mgr_;
 
-    std::string inertial_frame_name_, body_frame_name_, ned_frame_name_;
+    std::string inertial_frame_name_, body_frame_name_, ned_frame_name_,
+      earth_frame_name_, global_pos_sensor_frame_name_, local_pos_sensor_frame_name_;
+
+    cav_msgs::ExternalObject bsm_obj_;
```
Comment (Supraluminal) [inline sensor_fusion/src/sensor_fusion.h]: If this is just a singular `ExternalObject`  instance, does the BSM processing step work for multiple objects simultaneously still?

File: sensor_fusion/src/sensor_fusion.h [hunk 2]
```diff
@@ -118,7 +125,12 @@ class SensorFusionApplication
...
+
+    long unsigned int last_nav_sat_seq = -1;
```
Comment (Supraluminal) [inline sensor_fusion/src/sensor_fusion.h]: `unsigned` ... `= -1`, just wanted to check that it's intended for this to be MAX_LONG, and possibly architecture dependent?

File: sensor_fusion/src/wgs84_utils.cpp [hunk 3]
```diff
@@ -107,8 +107,73 @@ void wgs84_utils::convertToOdom(const wgs84_utils::wgs84_coordinate &src,
...
+ */
+//TODO lat/lon must be in radians
+tf2::Vector3 wgs84_utils::geodesic_2_cartesian(const wgs84_utils::wgs84_coordinate &loc, tf2::Transform ecef_in_ned) {
+
+    constexpr double Rea = 6378137.0; // Semi-major axis radius meters
+    constexpr double Rea_sqr = Rea*Rea;
+    constexpr double f = 1.0 / 298.257223563; //The flattening factor
+    constexpr double Reb = Rea * (1.0 - f); // //The semi-minor axis = 6356752.0
+    constexpr double Reb_sqr = Reb*Reb;
+    constexpr double e = 0.08181919084262149; // The first eccentricity (hard coded as optimization) calculated as Math.sqrt(Rea*Rea - Reb*Reb) / Rea;
```
[the remark this answers is not in the document]
Comment (msmcconnell) [reply sensor_fusion/src/wgs84_utils.cpp]: It seems std:sqrt is not a valid constexpr so I avoided using it directly.
https://stackoverflow.com/questions/8622256/in-c11-is-sqrt-defined-as-constexpr

File: sensor_fusion/src/wgs84_utils.cpp [hunk 4]
```diff
@@ -107,8 +107,73 @@ void wgs84_utils::convertToOdom(const wgs84_utils::wgs84_coordinate &src,
...
+    constexpr double e_sqr = e*e;
+    constexpr double e_p = 0.08209443794969568; // e prime (hard coded as optimization) calculated as Math.sqrt((Rea_sqr - Reb_sqr) / Reb_sqr);
```
[the remark this answers is not in the document]
Comment (msmcconnell) [reply sensor_fusion/src/wgs84_utils.cpp]: Ditto

File: sensor_fusion/src/sensor_fusion.h (hunk 1 above)
Comment (msmcconnell) [reply sensor_fusion/src/sensor_fusion.h]: That's just a left over object for debugging bsm fusion. I'll take it out. All the BSMs in the queue will be processed still.

File: sensor_fusion/src/transform_maintainer.h [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+#include <string>
+#include <unordered_map>
+#include <memory>
+#include <tf2_geometry_msgs/tf2_geometry_msgs.h>
+#include <cav_msgs/HeadingStamped.h>
+#include "wgs84_utils.h"
+
+namespace tf2
+{
+    inline // inline is required to make this compile. Not sure why
```
[the remark this answers is not in the document]
Comment (msmcconnell) [reply sensor_fusion/src/transform_maintainer.h]: I tracked this down.
It seems I was basing this off toMsg interface in tf2, but I added parameters. So it needed to be inline or it wouldn't overload. 
http://docs.ros.org/diamondback/api/tf2_geometry_msgs/html/namespacetf2.html
There were functions matching what I needed in tf2/buffer_core.cpp
http://docs.ros.org/api/tf2/html/namespacetf2.html#a9e63af3420b08fb187d1b9429a3fa899
But they are not exposed in the . So I have copied them over into the local tf2 namespace.  

The compilation issue matches what is described here
https://stackoverflow.com/questions/45667393/include-issue-multiple-definition-first-defined-here

I should have defined the functions in my cpp file

File: sensor_fusion/src/transform_maintainer.h [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+        msg.header.stamp = stamp;
+        msg.header.frame_id = frame_id;
+        msg.child_frame_id = child_frame_id;
+        
+        return msg;
+    }
+}
+
+/**
+ * TODO
```
[the remark this answers is not in the document]
Comment (msmcconnell) [reply sensor_fusion/src/transform_maintainer.h]: done

File: sensor_fusion/src/wgs84_utils.cpp [hunk 7]
```diff
@@ -107,8 +107,73 @@ void wgs84_utils::convertToOdom(const wgs84_utils::wgs84_coordinate &src,
...
+
+/**
+ * Converts a given WSG-84 geodesic location into a 3d cartesian point
+ * The returned 3d point is defined relative to a frame which has a transform with the ECEF frame.
+ * @param location The geodesic location to convert must be in radians
+ * @param frame2ecefTransform A transform which defines the location of the ECEF frame relative to the 3d point's frame of origin
+ * @return The calculated 3d point
+ */
+//TODO lat/lon must be in radians
+tf2::Vector3 wgs84_utils::geodesic_2_cartesian(const wgs84_utils::wgs84_coordinate &loc, tf2::Transform ecef_in_ned) {
```
[the remark this answers is not in the document]
Comment (msmcconnell) [reply sensor_fusion/src/wgs84_utils.cpp]: done

**Pull request merged**